querqy-core/src/main/java/querqy/rewrite/contrib/ReplaceRewriter.java

```diff
@@ -0,0 +1,111 @@
+package querqy.rewrite.contrib;
+
+import querqy.ComparableCharSequence;
+import querqy.CompoundCharSequence;
+import querqy.LowerCaseCharSequence;
+import querqy.model.AbstractNodeVisitor;
+import querqy.model.Clause;
+import querqy.model.DisjunctionMaxQuery;
+import querqy.model.ExpandedQuery;
+import querqy.model.Node;
+import querqy.model.Query;
+import querqy.model.Term;
+import querqy.rewrite.QueryRewriter;
+import querqy.trie.State;
+import querqy.trie.TrieMap;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import static querqy.rewrite.contrib.ReplaceRewriterParser.TOKEN_SEPARATOR;
+
+public class ReplaceRewriter extends AbstractNodeVisitor<Node> implements QueryRewriter {
+
+    private final TrieMap<List<ComparableCharSequence>> replaceRules;
+    private final boolean ignoreCase;
+
+    public ReplaceRewriter(TrieMap<List<ComparableCharSequence>> replaceRules, boolean ignoreCase) {
+        this.replaceRules = replaceRules;
+        this.ignoreCase = ignoreCase;
+    }
+
+    private LinkedList<ComparableCharSequence> querySeq = new LinkedList<>();
+    private LinkedList<ComparableCharSequence> matchSeq = new LinkedList<>();
+    private State<List<ComparableCharSequence>> priorMatch = new State<>(false, null, null);
+    private boolean hasReplacement = false;
+
+    @Override
+    public ExpandedQuery rewrite(ExpandedQuery query) {
+        querySeq.clear();
+        matchSeq.clear();
+
+        visit((Query) query.getUserQuery());
+
+        if (priorMatch.isFinal()) {
+            hasReplacement = true;
+            querySeq.addAll(priorMatch.value);
+        } else {
+            if (!matchSeq.isEmpty()) {
+                querySeq.addAll(matchSeq);
+            }
+        }
+
+        return hasReplacement ? buildQueryFromSeqList(querySeq) : query;
+    }
+
+    private ExpandedQuery buildQueryFromSeqList(LinkedList<ComparableCharSequence> tokens) {
+        Query query = new Query();
+        tokens.forEach(token -> {
+            DisjunctionMaxQuery dmq = new DisjunctionMaxQuery(query, Clause.Occur.SHOULD, false);
+            query.addClause(dmq);
+            Term term = new Term(dmq, token);
+            dmq.addClause(term);
+        });
+
+        return new ExpandedQuery(query);
+    }
+
+    @Override
+    public Node visit(final Term term) {
+
+        if (term.isGenerated()) {
+            return null;
```

Comment:
Arguably, we might also want to replace generated terms (could be a configuration setting) -> nice to have

Reply:
is the order of generated / ungenerated terms guaranteed? e. g. there is a query like Query(DMC(should, Term("phone", false))) and after common rules the query looks like Query(DMC(should, Term("phone", false)), DMC(should, Term("mobile", true)))

is it guaranteed that the order is always phone - mobile and never mobile - phone?

Reply:
i suggest that we ignore this for now and document the behavior of the rewriter precisely. if we have a better understanding about how this is supposed to work and about the use cases, we can add this later as an additional feature.

Reply:
Agreed, let's do this later. I'm also thinking about leaving more detailed information about the creator at the Terms / query objects, maybe some kind of rewriter name or 'purpose' ('commonrules/synonym' if the term/sub-query was created as a synonym)